### ISSUE TITLE
mip-nlp: a cold XLA compile (3m0.5s) blows the time limit on the OA path

### ISSUE THREAD

**Issue by jkitchin** [jkitchin/discopt#1063]
The `solver="mip-nlp"` path pulls JAX in and pays a cold XLA compile that is not
counted against, and vastly exceeds, the user's `time_limit`. On MINLPLib
`squfl015-060` with `time_limit=60`:

```
[Compiling module jit_concat_constraints for CPU] Very slow compile? ...
slow_operation_alarm.cc:140] The operation took 3m0.498402s

squfl015-060  oa  status=unknown  wall=199.7s      (time_limit was 60 s)
```

A single XLA compilation took **3 m 0.5 s** inside a 60 s budget. The run
returned at 199.7 s — over 3x the limit — having reached `status=unknown`.

Confirmed the path dependency on a fresh process:

```
jax after import discopt : False
jax after DEFAULT solve  : False
jax after OA solve       : True
```

So the default solve path is JAX-free as documented, and it is specifically the
MIP-NLP/OA route that imports JAX and triggers the compile.

Once the process is warm the compile is cached and the overrun is much smaller
(`squfl015-060` at `time_limit=20` returned in 20.1 s, 1.0x). The failure is
therefore a **cold-process** one: the first convex-MINLP solve in a fresh
interpreter can blow its budget by minutes. That is the case every batch
benchmark row and every short-lived worker actually hits.

A smaller, separate overrun remains after warm-up: `squfl020-150` at
`time_limit=20` returned in 28.4 s (1.4x), which is not explained by compilation
and should be triaged alongside.

## Why it matters now

#1059 proposes routing convexity-certified MINLPs to this path by default. A
3-minute cold-start compile must not become the default experience.

## What "done" means

- A cold-process convex-MINLP solve honours `time_limit` — either the JAX/XLA
  dependency is removed from this path (consistent with the rest of the solve
  stack, which is numpy + Rust), or the compile is cached/precompiled and the
  time spent is accounted against the budget rather than added to it.
- Regression test asserting wall <= time_limit * (1 + slack) on a cold
  interpreter for a convex MINLP.
- The residual 1.4x warm overrun on `squfl020-150` is explained.

## Measurement conditions

14-core machine, elevated load (avg 5-10) per CLAUDE.md §9, so the 199.7 s
figure is inflated in magnitude; the 3 m 0.5 s compile is reported by XLA's own
alarm and the 3x-over-budget character of the failure is not a load artifact.

**Comment by jkitchin** [comment]
Tracked by #1066, which carries the full analysis, the measurement conditions, and the SCIP/BARON comparison this was drawn from.

**Comment by jkitchin** [comment]
## Root cause pinned: a bypassed evaluator dispatcher, not a missing XLA cache

The framing in the issue body ("worth precompiling/caching") is wrong, and I am
retracting it here per CLAUDE.md §11. There is nothing to precompile: **JAX
should not be on this path at all**, and it is there because one call site
reaches past the dispatcher that exists to keep it out.

`discopt._tape_nlp_evaluator.make_evaluator()` is the single decision point
between the POUNCE tape evaluator (numpy + Rust, default-ON since the §5 panel)
and the legacy JAX `NLPEvaluator`. Its docstring states the invariant:

> The `cached_evaluator` import is inside the fallback on purpose:
> `_relax/nlp_evaluator` imports jax at module scope, so importing it eagerly
> puts JAX in `sys.modules` on every nonlinear solve and defeats the tape
> backend entirely (#75). **Callers must not reach past this to
> `cached_evaluator` — that is what kept JAX on the path after Stage 3.**

The OA path reaches past it:

```
python/discopt/solvers/oa.py:1183:    evaluator = NLPEvaluator(model)      # in _decompose_model
```

Chain: `mip_nlp.py:895 solve_oa` → `oa.py:_decompose_model` →
`NLPEvaluator(model)` → `_relax/nlp_evaluator.py:22 import jax` →
`_relax/nlp_evaluator.py:391 self._cons_fn_jit = jax.jit(concat_constraints)`.

That last line is exactly the `jit_concat_constraints` module XLA's slow-compile
alarm named. So the 3 m 0.5 s is a first-trace XLA compile of a graph that the
tape backend would never have built.

This also explains the measured `jax after DEFAULT solve: False` /
`jax after OA solve: True` asymmetry cleanly: the default path goes through
`make_evaluator`, the OA path does not.

## Other direct constructors on reachable solve paths

Same bypass, same class of defect (`solvers/` and orchestration, excluding the
`_relax/differentiable*.py` and `_relax/pounce_layer.py` sites, which are the
differentiable-solve subsystem and are JAX by design):

```
python/discopt/solvers/oa.py:1183
python/discopt/solvers/gdpopt_loa.py:69
python/discopt/solvers/amp.py:1029, 2440
python/discopt/solvers/_convex_kernel.py:485
python/discopt/solvers/sipopt.py:180, 211, 218
python/discopt/decomposition/benders/gbd.py:323
python/discopt/validation/examiner.py:170
python/discopt/mpec.py:358
python/discopt/warm_start.py:254
```

Per CLAUDE.md §2 the fix is the class, not `oa.py`: route these through
`make_evaluator` so the tape backend is used wherever the model is
tape-representable, with the JAX evaluator retained as the documented fallback
for unrepresentable models (array-valued bodies, `dm.custom`, operators with no
tape lowering).

## Revised "done"

- No `solvers/`-layer or orchestration call site constructs `NLPEvaluator`
  directly; all go through `_tape_nlp_evaluator.make_evaluator`.
- Regression test: a cold interpreter solving a convex MINLP via
  `solver="mip-nlp"` ends with `"jax" not in sys.modules` (an
  anti-vacuity control asserting the solve actually ran, per §6).
- Cold-process wall honours `time_limit`.
- The residual 1.4x warm overrun on `squfl020-150` is still to be explained
  separately — it is not a compile artifact.
